src/dune_engine/build_config_intf.ml

```diff
@@ -13,13 +13,20 @@ end
 
 type rules = Rules.t
 
+module Directory_targets = struct
+  type t =
+    { all : Loc.t Path.Build.Map.t Memo.t
+    ; mem : Path.Build.t -> bool Memo.t
+    }
+end
```

Comment:
Since this type has invariants I think it would be better to not allow clients to constructs instances which break them. So it might be better to keep it abstract and expose `all` and `mem` as functions.